libass/ass_fontconfig.c

```diff
@@ -191,6 +202,11 @@ static void scan_fonts(FcConfig *config, ASS_FontProvider *provider)
 
         ass_font_provider_add_font(provider, &meta, path, index, (void *)pat);
     }
+
+    // FcFontSort's, but not FcConfigGetFOnt's!, FOntSet needs to be cleaned up explicitly
+    if (res == FcResultMatch) {
+        FcFontSetDestroy(fonts);
```

Comment:
Digging into this, I'm arriving at the conclusion that this code is currently technically safe, but only by chance: `FcFontSetDestroy` unrefs all the `FcPattern`s that we've just inserted into our font list. All of those patterns ultimately come straight out of our `FcConfig`, which will hold a ref to them until we exit, but I don't think there's any guarantee that that'll always be true (and it's also kinda fragile even in the existing code). Instead, we should either:
- skip destroying `fonts` here and instead hold onto it in `ProviderPrivate` until `destroy` is called, or
- `FcPatternReference` each pattern before calling `ass_font_provider_add_font`, and introduce a `destroy_font` callback that releases it

I'm slightly partial to option 2, since it'd lay some nice groundwork for a lazy-loading fontconfig provider future, and ensures we're not relying on any potentially-non-guaranteed lifetimes in the `FcConfigGetFonts` fallback case.

Reply:
very good point; implemented referencing and `destroy_font`